Comment on that all nonlinear operators can be shifted to augmented primal (#587)
JuliaDiff/ChainRulesCore.jl: docs/src/design/changing_the_primal.md

We have been promised them for Julia v1.7 though.
You can see what the code would look like in [PR #302](https://github.com/JuliaDiff/ChainRules.jl/pull/302).

> ## What things can be pulled out of the pullback?

Reviewer: ```suggestion
## What things can be taken out of the pullback?
```
Seems clearer?

Reviewer: My understanding is that the motivation is that we can reuse the work that was done in the primal to reduce the work that needs to be done in the pullback.

But this current paragraph insinuates (at least to me) that: `if there is an operation you can do in the augmented primal, do it there, rather than in the pullback.` Is this true? I can imagine this is true if pullback gets called more than once, but that does not happen, right?

Reviewer 2: The pullback gets called several times by `jacobian`.

It also sometimes does not get called at all, which should happen when the gradient is Zero (but doesn't always?) and may also happen because AD has called `rrule` on code it ought to know cannot have a derivative (as in https://github.com/FluxML/NNlib.jl/pull/434 ).

At this point you might wonder, is there a rule for what can be taken out of the pullback and computed in the augmented primal?
We can deduce one, or infact two.

> The first and most practical one is that any computation that depends only on the primal input (or consequently its output) can be shifted to the augmented primal.

Reviewer: ```suggestion
The first and most practical one is that any computation that depends only on the primal input or output can be shifted to the augmented primal.
```

> The second and perhaps more insightful is that all nonlinear parts can moved out (this is a weaker statement but more interesting).

Reviewer: ```suggestion
The second, perhaps more insightful, rule is that all nonlinear parts can be moved out.
This is a weaker but more interesting statement.
```

> We know this because pullbacks are linear operators -- they are linear in relation to the tangent they are pulling back.

Reviewer: ```suggestion
We know this because pullbacks are linear operators -- linear w.r.t. the tangent they are pulling back.
```

> This means they are in turn composed only of functions that linear operators (in relation to the tangent).

Reviewer: ```suggestion
This means they are in turn composed only of functions that are linear operators w.r.t. the tangent.
```

> The fully minimized pullback function only calls linear operators -- the apparently nonlinear parts can all be shifted to the augmented primal.

Reviewer: ```suggestion
The minimal pullback function only calls linear operators -- the apparently nonlinear parts can all be shifted to the augmented primal.
```

Not sure what "the apparently nonlinear parts" means though?

Author: idk what the apparently was about either
```suggestion
The minimal pullback function only calls linear operators -- the nonlinear parts can all be shifted to the augmented primal.
```

We have been promised them for Julia v1.7 though.
You can see what the code would look like in [PR #302](https://github.com/JuliaDiff/ChainRules.jl/pull/302).

## What things can be taken out of the pullback?
At this point you might wonder, is there a rule for what can be taken out of the pullback and computed in the augmented primal?

> We can deduce one, or infact two.

Reviewer 3: ```suggestion
We can deduce one, or in fact two.
```
except if I just didn't know the one-word spelling
